macros/contextFraction.pl

```diff
@@ -627,15 +627,17 @@ sub string {
 }
 
 #
-#  Add parentheses if they are needed by precedence
+#  Add parentheses if they were there originally, or
+#  are needed by precedence and we asked for exxxtra parens
 #
 sub TeX {
   my $self = shift;
-  my $string = $self->SUPER::TeX($self, @_);
+  my $string = $self->SUPER::TeX(@_);
   return $string unless $self->{value}->classMatch('Fraction');
   my $precedence = shift;
   my $frac = $self->context->operators->get('/')->{precedence};
-  $string = '\left(' . $string . '\right)' if defined $precedence && $precedence > $frac;
+  $string = '\left(' . $string . '\right)' if $self->{hadParens} ||
+    (defined $precedence && $precedence > $frac && $self->context->flag('showExtraParens') > 1);
```

Comment:
Adding the test for `showExtraParens` is not the right solution to the unary minus issue, as these parens are not optional for other situations.  For example, if `showExtraParens => 0` is set, then something like

```
(Formula(Fraction(1,2))**2)->TeX
```

will produce a fraction squared, but without the needed parentheses.  That is not the correct display for that situation.

Fortunately, the unary minus already passes a parameter to the `TeX()` method to indicate that there should be no parens around a fraction, and you can tap into that.  It is the next parameter in the argument list, so using

```suggestion
  my $noparens = shift;
  $string = '\left(' . $string . '\right)' if $self->{hadParens} ||
    (defined $precedence && $precedence > $frac && !$noparens);
```

should make it work both for unary minus and powers (or other higher-precedence operations).

Reply:
Yes, that is right on both counts.

Reply:
The `string` and `TeX` methods are a mess, and probably deserve an overhaul.  Lots of hacks in there to make it work that should probably be cleaned up.

Reply:
also, you have a mixture of my thanks and apologies for belaboring all of this -- I've benefited greatly from hacking away on this issue, but I'm also super grateful for anything you're willing to throw my way in terms of making sense of the overall structure. 🙏

Reply:
Well, this takes a bit of history and overview to understand.

First, you need to understand the difference between the `Parser` namespace and the `Value` namespace.  The `Parser` objects are used to maintain an abstract syntax tree for a parsed formula, so the just represent the structure of the formula.  Thus things like addition and powers and function calls and how they fit together are coded here. In general, this is not about computation, but about the structure.

The `Value` objects are the ones that model the actual mathematical values of things like real and complex numbers, vectors and such, and do the computations involving them.  There is no structure here, just computation.  So if a real number cam from addition of two numbers or multiplication is not known here.  We just have the number, not how it was computed.

Originally, I wanted these two libraries to be able to operate separately, in particular the Value objects were supposed to be able to be used without the Parser, just for computational purposes.  As it turns out, over time they became more and more intertwined, and the separation is no longer possible.

The `Value::Formula` is also a complication, since it is really a `Parser` object under the hood, but it has all the computational methods for doing things like evaluating the formula, generating random points, comparing formulas and so on.  So it is a cross-over object.

A complication from the other direction is that the Parser can reduce constant operations and function calls, and so when it parses `1 + 2`, rather than making the abstract tree with a `Parser::BOP::add` node whose `lop` and `rop` fields are `Parser::Number` objects whose `value` fields are the Perl reals 1 and 2, the parser converts the perl reals into `Value::Real` objects and asks one to add itself to the other, and gets back the `Value::Real` result.  But this doesn't fit into a `Parser::Number` since its value is a perl real.  In the case of a `Value:Real()`, the parser knows how to convert the result back into a perl number for the `Parser::Number`, but in general, the parser doesn't really know much about `Value` objects or how they relate to its own objects.  The `Value` object that the parser gets back from reducing constants or function calls could be virtually anything, like something defined in an external macro file, for example a Fraction object, which the parser has never seen before.  That means the parser needs another mechanism of dealing with such results since it can't turn them back into its own parser nodes.  That is handled by the `Parser::Value` object, which is just a shell around a `Value` object as a holding place for it (whether it is a vector, list, or some subclass that is unknown to it, like a fraction).

These `Parser::Value` nodes are only created if the parser has to deal with a `Value` object (other than a Real or Complex object).  That can happen due to reducing constants, as in the example above, or by doing something like passing a `Value` object to the `Formula()` constructor (which will create a formula containing just a `Parser::Value` node holding the `Value` object.  So `Formula(Vector(1,2,3))` is different from doing `Formula("<1,2,3>")` (when `reduceConstants => 0` is set) since the former creates a `Parser::Value` node containing a `Value::Vector` object, while the later produces a `Parser::list::vector` node with three `Parser::Number` sub-nodes.

Now the context can override the objects that the Parser creates in several ways.  For example, you can set the class created by an operator in the context's operators hash.  For example, the `contextFraction.pl` file sets up the `Fraction` context using 

``` perl
$context->operators->set("/"  => {class => "context::Fraction::BOP::divide"});
```

(along with some other operator definitions).  This overrides the `/` operator to produce a `Fraction::BOP::divide` object rather than the usual `Parser::BOP::divide` that is the default for contexts based on the Numeric context.  When a `parser::BOP::divide` node is evaluated, it generates a `Fraction` object rather than a `Real`.  (Technically, it is a `context::Fraction::Fraction` object, which is the `Value` object that handles computation with fractions.)  This is how `Compute("1/2")` is made to produce a Fraction object in the end.

The context can also override the class used to create the various `Parser` nodes or `Value` objects that it needs.  For example, the Fraction context also sets

``` perl
$context->{value}{Real} = "context::Fraction::Real";
$context->{parser}{Value} = "context::Fraction::Value";
```

and some other definitions.  This tells the parser that when it wants to create a `Value::Real` it should instead create a `context::Fraction::Real` and when it wants to create a `Parser::Value` node, it should create `context::Fraction::Value` instead.  This means that if you do `Formula(Fraction(1,2))` (or `Formula("1/2")` when `reduceConstants => 1` is set), you should get a `Fraction::Value` node rather than a `Parser::Value` node.  The `Fraction::Value` node is what allows constant Fraction values to be handled specially (in particular for its `string` and `TeX` methods).

So we are finally in a position to answer your question about `Parser::UOP->{op}`.  When the `UOP` is acting on a fraction, it could be either a `Fraction::BOP::divide` or a `Fraction::Value` containing a `Fraction::Fraction`  object, depending on how the the operator was created.  If the fraction was a constant fraction and `reduceConstants` is true, the `Parser::UOP->{op}` will have a `Fraction::Value` object, and when its `TeX()` method is called, it will be the one you have been modifying in the `Fraction::Value` class.  That will call the `Fraction::Value` superclass (which is `Parser::Value`) `TeX()` method, and that in turn will call the `Value` object's `TeX()` method.  In this case, the `Value` object is a `Fraction::Fraction`, and so that is the `TeX()` method that will be called.  These are the ones that used to contain the `($n)` lines that you removed.

On the other hand, if the UOP is acting on a fraction that is not constant (e.g., `1/x`) or `reduceConstants` is false, then the `Parser::UOP->{op}` will be a `Fraction::BOP::divide`, and it is that object's `TeX()` method that is called.  That is a subclass of `Parser::BOP::divide`, and it overrides `TeX()` but not `string()`.  When its `TeX()` method is called, it calls `SUPER::TeX()`, so *that* is when `Parser::BOP::divide->TeX()` is finally called.  That will only happen if the fraction is not constant, or `reduceConstants` is false, and the parser tree holds the `Fraction::BOP::divide` and its two `Number` nodes, rather than the `Fraction::Value` node.

You can investigate the parse tree using `TEXT(Formula("...")->{tree})` to print the reference to the top node in the tree, and you could `pretty_print()` the tree to see the values of its fields, I would think.

Anyway, that is a long-winded response, but I hope it helps clarify what is happening, and why you have been seeing different values in the output for the table in the original issue thread.



Similarly, when the Fraction context is used (with `reduceConstants => 0`), `Formula(Fraction(1,2))` produces a `Parser::Value`